src/metabase/api/alert.clj

```diff
@@ -0,0 +1,182 @@
+                        (contains? (recipient-ids alert) api/*current-user-id*)))))
+
+(defn- email-channel [alert]
+  (m/find-first #(= :email (:channel_type %)) (:channels alert)))
+
+(defn- slack-channel [alert]
+  (m/find-first #(= :slack (:channel_type %)) (:channels alert)))
+
+(defn- key-by [key-fn coll]
+  (zipmap (map key-fn coll) coll))
+
+(defn- notify-recipient-changes!
+  "This function compares `OLD-ALERT` and `UPDATED-ALERT` to determine if there have been any recipient related
+  changes. Recipients that have been added or removed will be notified."
+  [old-alert updated-alert]
+  (let [{old-recipients :recipients} (email-channel old-alert)
+        {new-recipients :recipients} (email-channel updated-alert)
+        old-ids->users (key-by :id old-recipients)
+        new-ids->users (key-by :id new-recipients)
+        [removed-ids added-ids _] (data/diff (set (keys old-ids->users))
+                                             (set (keys new-ids->users)))]
+    (doseq [old-id removed-ids
+            :let [removed-user (get old-ids->users old-id)]]
+      (messages/send-admin-unsubscribed-alert-email! old-alert removed-user @api/*current-user*))
+
+    (doseq [new-id added-ids
+            :let [added-user (get new-ids->users new-id)]]
+      (messages/send-you-were-added-alert-email! updated-alert added-user @api/*current-user*))))
+
+(api/defendpoint PUT "/:id"
+  "Update a `Alert` with ID."
+  [id :as {{:keys [alert_condition card channels alert_first_only alert_above_goal card channels] :as req} :body}]
+  {alert_condition  AlertConditions
+   alert_first_only s/Bool
+   alert_above_goal (s/maybe s/Bool)
+   card             su/Map
+   channels         (su/non-empty [su/Map])}
+  (let [old-alert     (pulse/retrieve-alert id)
+        _             (check-alert-update-permissions old-alert)
+        updated-alert (-> req
+                          only-alert-keys
+                          (assoc :id id :card (u/get-id card) :channels channels)
+                          pulse/update-alert!)]
+
+    ;; Only admins can update recipients
+    (when (and api/*is-superuser?* (email/email-configured?))
+      (notify-recipient-changes! old-alert updated-alert))
+
+    updated-alert))
+
+(defn- should-unsubscribe-delete?
+  "An alert should be deleted instead of unsubscribing if
+     - the unsubscriber is the creator
+     - they are the only recipient
+     - there is no slack channel"
+  [alert unsubscribing-user-id]
+  (let [{:keys [recipients]} (email-channel alert)]
+    (and (= unsubscribing-user-id (:creator_id alert))
+         (= 1 (count recipients))
+         (= unsubscribing-user-id (:id (first recipients)))
+         (nil? (slack-channel alert)))))
+
+(api/defendpoint PUT "/:id/unsubscribe"
+  [id]
+  ;; Admins are not allowed to unsubscribe from alerts, they should edit the alert
+  (api/check (not api/*is-superuser?*)
+    [400 "Admin user are not allowed to unsubscribe from alerts"])
+  (assert (integer? id))
```

Comment:
defendpoint already adds regexes for `:id` forcing them to be integer strings and then takes care of parsing them for you so there's no way you could get here and have `id` not be an integer. At any rate if you want to validate params in an endpoint you should use the `{param schema}` map because those get picked up and added to the auto-generated documentation and the produce errors in a format understandable by the frontend, unlike `assert`